########################################################################################################################
# Hexapod Parameters
########################################################################################################################

hexapod:
  parameters:
  
########################################################################################################################
    # Control parameters
    time_delta: 0.02
    imu_compensation: false #requires imu
    auto_compensation: true
    manual_compensation: true
    inclination_compensation: false #requires imu
    impedance_control: false
    imu_rotation_offset: [3.14, 0.0, 0.0]
    interface_setup_speed: 0.5
    
########################################################################################################################
    # Model parameters
    hexapod_type: max
    leg_id: [AR, BR, CR, CL, BL, AL] # leg_id_name (clockwise from front right leg)
    joint_id: [coxa, femur, tibia] # joint_id_name
    link_id: [base, coxa, femur, tibia] # link_id_name (always start with base) 
    leg_DOF: {AL: 3, AR: 3, BL: 3, BR: 3, CL: 3, CR: 3} # leg_id_name: degrees_of_freedom (num joints) 
    leg_stance_yaws: {AL: 0.785, AR: -0.785, BL: 1.57, BR: -1.57, CL: 2.355, CR: -2.355}
    # Joint parameters:
    AR_coxa_joint_parameters:  {offset: 1.30,   min: -1.30,  max: 0.270, packed: 1.57,  unpacked: 0.0,   max_vel: 1.00}
    AR_femur_joint_parameters: {offset: 0.0,    min: -1.30,  max: 0.436, packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    AR_tibia_joint_parameters: {offset: 0.0,    min: 0.872,  max: 2.79,  packed: -1.57, unpacked: 0.785, max_vel: 0.40}    
    BR_coxa_joint_parameters:  {offset: 0.0,    min: -0.785, max: 0.785, packed: -1.57, unpacked: 0.0,   max_vel: 1.00}
    BR_femur_joint_parameters: {offset: 0.0,    min: -1.30,  max: 0.436, packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    BR_tibia_joint_parameters: {offset: 0.0,    min: 0.872,  max: 2.79,  packed: 1.57,  unpacked: 0.785, max_vel: 0.40}    
    CR_coxa_joint_parameters:  {offset: -1.30,  min: -0.270, max: 1.30,  packed: -1.57, unpacked: 0.0,   max_vel: 1.00}
    CR_femur_joint_parameters: {offset: 0.0,    min: -1.30,  max: 0.436, packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    CR_tibia_joint_parameters: {offset: 0.0,    min: 0.872,  max: 2.79,  packed: -1.57, unpacked: 0.785, max_vel: 0.40}
    AL_coxa_joint_parameters:  {offset: -1.30,  min: -0.270, max: 1.30,  packed: 1.57,  unpacked: 0.0,   max_vel: 1.00}
    AL_femur_joint_parameters: {offset: 0.0,    min: -0.436, max: 1.30,  packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    AL_tibia_joint_parameters: {offset: 0.0,    min: -2.79,  max: -0.872,packed: -1.57, unpacked: 0.785, max_vel: 0.40}
    BL_coxa_joint_parameters:  {offset: 0.0,    min: -0.785, max: 0.785, packed: -1.57, unpacked: 0.0,   max_vel: 1.00}
    BL_femur_joint_parameters: {offset: 0.0,    min: -0.436, max: 1.30,  packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    BL_tibia_joint_parameters: {offset: 0.0,    min: -2.79,  max: -0.872,packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    CL_coxa_joint_parameters:  {offset: 1.30,   min: -1.30,  max: 0.270, packed: -1.57, unpacked: 0.0,   max_vel: 1.00}
    CL_femur_joint_parameters: {offset: 0.0,    min: -0.436, max: 1.30,  packed: 1.57,  unpacked: 0.785, max_vel: 0.40}
    CL_tibia_joint_parameters: {offset: 0.0,    min: -2.79,  max: -0.872,packed: -1.57, unpacked: 0.785, max_vel: 0.40}
    # Link DH parameters: en.wikipedia.org/wiki/Denavitâ€“Hartenberg_parameters    
    AR_base_link_parameters:  {d: 0.0, theta: -0.270, r: 0.589, alpha: 0.0}
    AR_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: -1.57}
    AR_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    AR_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}    
    BR_base_link_parameters:  {d: 0.0, theta: -1.57,  r: 0.158, alpha: 0.0}
    BR_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: -1.57}
    BR_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    BR_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}    
    CR_base_link_parameters:  {d: 0.0, theta: -2.87,  r: 0.589, alpha: 0.0}
    CR_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: -1.57}
    CR_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    CR_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}
    AL_base_link_parameters:  {d: 0.0, theta: 0.270,  r: 0.589, alpha: 0.0}
    AL_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: 1.57}
    AL_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    AL_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}
    BL_base_link_parameters:  {d: 0.0, theta: 1.57,   r: 0.158, alpha: 0.0}
    BL_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: 1.57}
    BL_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    BL_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}
    CL_base_link_parameters:  {d: 0.0, theta: 2.87,   r: 0.589, alpha: 0.0}
    CL_coxa_link_parameters:  {d: 0.0, theta: 0.0,    r: 0.080, alpha: 1.57}
    CL_femur_link_parameters: {d: 0.0, theta: 0.0,    r: 0.800, alpha: 0.0}
    CL_tibia_link_parameters: {d: 0.0, theta: 0.872,  r: 1.50,  alpha: 0.0}
    
########################################################################################################################
    # Walker parameters
    gait_type: wave_gait
    step_frequency: {default: 0.09, min: 0.001, max: 0.200, step: 0.01} #Reconfigurable
    step_clearance: {default: 0.15, min: 0.05, max: 1.0, step: 0.01} #Reconfigurable
    step_depth: 0.025 #0.025
    body_clearance: {default: 1.0, min: 0.5, max: 2.0, step: 0.1} #Reconfigurable
    leg_span: {default: 1.0, min: 0.7, max: 0.8, step: 0.01} #Reconfigurable
    velocity_input_mode: throttle #real    
    force_cruise_velocity: true
    linear_cruise_velocity: {x: 1.0, y: 0.0}
    angular_cruise_velocity: 0.0

########################################################################################################################
    # Poser parameters
    start_up_sequence: false #DO NOT USE
    time_to_start: 12.0
    rotation_pid_gains: {p: 0.0, i: 0.0, d: 0.0}
    translation_pid_gains: {p: 0.0, i: 0.0, d: 0.0}
    auto_compensation_parameters: {pitch: 0.02, roll: 0.015, z_trans: 0.05}
    max_translation: {x: 0.35, y: 0.5, z: 0.5}
    max_rotation: {roll: 0.075, pitch: 0.15, yaw: 0.1}
    max_translation_velocity: 0.05
    max_rotation_velocity: 0.05
    leg_manipulation_mode: tip_control

########################################################################################################################
    # Impedance controller parameters
    dynamic_stiffness: true
    use_joint_effort: false
    virtual_mass: {default: 100, min: 10, max: 1000, step: 10} #Reconfigurable
    integrator_step_time: 0.075
    virtual_stiffness: {default: 50, min: 1, max: 1000, step: 10} #Reconfigurable
    load_stiffness_scaler: 10.0
    swing_stiffness_scaler: 0.02
    virtual_damping_ratio: {default: 4.0, min: 0.1, max: 10.0, step: 0.05} #Reconfigurable
    force_gain: {default: 0.05, min: 0.001, max: 100, step: 1.0} #Reconfigurable
    
########################################################################################################################
    # Debug parameters
    console_verbosity: debug
    debug_move_to_joint_position: false
    debug_step_to_position: false
    debug_swing_trajectory: false
    debug_stance_trajectory: false
    debug_ik: false
    debug_rviz: true
    debug_rviz_static_display: true
    
########################################################################################################################
########################################################################################################################
